src/cascadia/TerminalSettingsModel/ActionMap.cpp

```diff
@@ -59,31 +59,83 @@ namespace winrt::Microsoft::Terminal::Settings::Model::implementation
     }
 
     // Method Description:
-    // - Retrieves the Command in the current layer, if it's valid
-    // - We internally store invalid commands as full commands.
-    //    This helper function returns nullptr when we get an invalid
-    //    command. This allows us to simply check for null when we
-    //    want a valid command.
-    // Arguments:
-    // - actionID: the internal ID associated with a Command
-    // Return Value:
-    // - If the command is valid, the command itself.
-    // - If the command is explicitly unbound, nullptr.
-    // - If the command cannot be found in this layer, nullopt.
-    std::optional<Model::Command> ActionMap::_GetActionByID(const InternalActionID actionID) const
-    {
-        // Check the masking actions
-        const auto maskingPair{ _MaskingActions.find(actionID) };
-        if (maskingPair != _MaskingActions.end())
+    // - Detects if any of the user's actions are identical to the inbox actions,
+    //   and if so, deletes them and redirects their keybindings to the inbox actions
+    // - We have to do this here instead of when loading since we don't actually have
+    //   any parents while loading the user settings, the parents are added after
+    void ActionMap::_FinalizeInheritance()
+    {
+        // first, gather the inbox actions from the relevant parent
+        std::unordered_map<InternalActionID, Model::Command> inboxActions;
+        winrt::com_ptr<implementation::ActionMap> foundParent{ nullptr };
+        for (const auto& parent : _parents)
         {
-            // ActionMap should never point to nullptr
-            FAIL_FAST_IF_NULL(maskingPair->second);
+            for (const auto& [_, cmd] : parent->_ActionMap)
+            {
+                if (cmd.Origin() != OriginTag::InBox)
+                {
+                    // only one parent contains all the inbox actions and that parent contains only inbox actions,
+                    // so if we found a non-inbox action we can just skip to the next parent
+                    continue;
+                }
+                foundParent = parent;
+                break;
+            }
+        }
 
-            // masking actions cannot contain nested or invalid commands,
-            // so we can just return it directly.
-            return maskingPair->second;
+        if (foundParent)
+        {
+            for (const auto& [_, cmd] : foundParent->_ActionMap)
+            {
+                inboxActions.emplace(Hash(cmd.ActionAndArgs()), cmd);

```

Comment:
can you just combine this with the upper loops?

Reply:
I wanted to avoid repeatedly checking for `Origin() != InBox` once we had already determined the correct parent! Also to avoid repeatedly setting `foundParent = parent`

Reply:
yep! that’d be why you break out of all loops after you find the parent and populate the content.